activesupport/lib/active_support/messages/metadata.rb

```diff
@@ -37,7 +37,11 @@ def pick_expiry(expires_at, expires_in)
           end
 
           def extract_metadata(message)
-            data = JSON.decode(message) rescue nil
+            # Rescuing an exception here can be wildly expensive here. We noticed 22ms overhead
+            # on each request when attempting to decode metadata that was serialized using Marshal.
+            #
+            # Real fix here is to have `Metadata` aware of the serializer.
+            data = JSON.decode(message) if message&.first == "{"
```

Comment:
Just to mention, other values are also valid for `JSON.decode`. For example numbers.

```
ActiveSupport::JSON.decode('5')
# => 5
```

Reply:
`JSON.decode` here is only to [extract metadata](https://github.com/rails/rails/blob/a3e392f65614a744ca0f2b509301538e2a28bb95/activesupport/lib/active_support/messages/metadata.rb#L42).  If a message has metadata, it will start with `"{"`, but if it doesn't have metadata, it will be [handled separately](https://github.com/rails/rails/blob/a3e392f65614a744ca0f2b509301538e2a28bb95/activesupport/lib/active_support/messages/metadata.rb#L45) (note that line is `new(message)` rather than `new(data)`).